src/consensus/softforks.cpp

```diff
@@ -0,0 +1,144 @@
+// Copyright (c) 2015 Eric Lombrozo
+// Copyright (c) 2009-2015 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include "consensus/softforks.h"
+
+#include "chain.h"
+#include "consensus/blockruleindex.h"
+#include "consensus/params.h"
+
+using namespace Consensus::SoftForks;
+
+/**
+ * Returns true if there are nRequired or more blocks of minVersion or above
+ * in the last Consensus::Params::nMajorityWindow blocks, starting at pstart and going backwards.
+ */
+static bool IsSuperMajority(int minVersion, const CBlockIndex* pstart, unsigned nRequired, const Consensus::Params& consensusParams)
+{
+    unsigned int nFound = 0;
+    for (int i = 0; i < consensusParams.nMajorityWindow && nFound < nRequired && pstart != NULL; i++)
+    {
+        if (pstart->nVersion >= minVersion)
+            ++nFound;
+        pstart = pstart->pprev;
+    }
+    return (nFound >= nRequired);
+}
+
+VersionStatus Consensus::SoftForks::CheckVersion(const CBlockIndex& blockIndex, const Consensus::VersionBits::BlockRuleIndex& blockRuleIndex, const Consensus::Params& consensusParams, CBlockIndex* pindexPrev)
+{
+    using namespace Consensus::VersionBits;
+
+    if (!pindexPrev)
+        pindexPrev = blockIndex.pprev;
+
+    // Handle versionbits cases first
+    if (UsesVersionBits(blockIndex.nVersion))
+        return blockRuleIndex.AreVersionBitsRecognized(&blockIndex, pindexPrev) ? VALID : UNRECOGNIZED;
+
+    // Reject blockIndex.nVersion=1 blocks when 95% (75% on testnet) of the network has upgraded:
+    if (blockIndex.nVersion < 2 && IsSuperMajority(2, pindexPrev, consensusParams.nMajorityRejectBlockOutdated, consensusParams))
+        return INVALID;
+
+    // Reject blockIndex.nVersion=2 blocks when 95% (75% on testnet) of the network has upgraded:
+    if (blockIndex.nVersion < 3 && IsSuperMajority(3, pindexPrev, consensusParams.nMajorityRejectBlockOutdated, consensusParams))
+        return INVALID;
+
+    // Reject blockIndex.nVersion=3 blocks when 95% (75% on testnet) of the network has upgraded:
+    // DEPLOY BIP65 - Uncomment the following line to deploy
+    //if (blockIndex.nVersion < 4 && IsSuperMajority(4, pindexPrev, consensusParams.nMajorityRejectBlockOutdated, consensusParams))
+    //    return false;
+
+    return VALID;
+}
+
+bool Consensus::SoftForks::UseRule(int rule, const CBlockIndex& blockIndex, const Consensus::VersionBits::BlockRuleIndex& blockRuleIndex, const Consensus::Params& consensusParams, CBlockIndex* pindexPrev)
```

Comment:
Why ```int rule``` instead of Consensus::SoftForks::Rule ? This happens in many functions/methods, in class attributes and with other enums like RuleState.